File: jogo_dos_8/src/game/algoritmos/BuscaEmLargura.java
package game.algoritmos;

import java.util.ArrayList;
import java.util.List;

import game.Tabuleiro;
import game.Movimento;

public class BuscaEmLargura {
	private List<Tabuleiro> fronteira = new ArrayList<Tabuleiro>();
	private List<Tabuleiro> visitado = new ArrayList<Tabuleiro>();
	
	public BuscaEmLargura(Tabuleiro jogo) {
		fronteira.add(jogo);
		execute();
	}
	
	private boolean execute() {
		int contador_Seguro = 0;
		while(true) {
			Tabuleiro t = fronteira.get(0);
			fronteira.remove(0);
			visitado.add(t);
			if(t.isSolucao()) {
				System.err.println(t);
				return true;
			}
			
			int[] p = Tabuleiro.getLocalizacaoDoVazio(t);
			int x = p[1];
			int y = p[0];
			System.out.println("x: " + x + " y: " + y);
			System.out.println(t);
			
			if(Movimento.podeMoverCima(p)) {
				Tabuleiro novo = Movimento.moverParaCima(t, p);
				if(!visitado.contains(novo)) {
					fronteira.add(novo);
					visitado.add(novo);
				}
			}
			if(Movimento.podeMoverEsq(p)) {
				Tabuleiro novo = new Tabuleiro();
				novo = Movimento.moverParaEsquerda(t, p);
				if(!visitado.contains(novo)) {
					fronteira.add(novo);
					visitado.add(novo);
				}
			}
			if(Movimento.podeMoverDir(p, t.getSize())) {
				Tabuleiro novo = new Tabuleiro();
				novo = Movimento.moverParaDireita(t, p);
				if(!visitado.contains(novo)) {
					fronteira.add(novo);
					visitado.add(novo);
				}
			}
			if(Movimento.podeMoverBaixo(p, t.getSize())) {
				Tabuleiro novo = Movimento.moverParaBaixo(t, p);
				if(!visitado.contains(novo)) {
					fronteira.add(novo);
					visitado.add(novo);
				}
			}
			
			if(contador_Seguro > 1000) {
				break;
			}
			contador_Seguro++;
		}
		
		return true;
	}
	
	public static void main(String[] args) {
//		Tabuleiro a = new Tabuleiro(3);
		//System.out.println(a);
		
		int [][] sol = {{1,2,3},{0,4,5},{7,8,6}};
		Tabuleiro solucionavel = new Tabuleiro(sol);
		new BuscaEmLargura(solucionavel);
	}
}

File: algoritmos/BuscaEmLargura.py

from game.Movimento import Movimento
from game.Tabuleiro import Tabuleiro
from queue import Queue
import numpy as np
import random


class BuscaEmLargura:
    def __init__(self):
        self.fronteira = Queue()

    def execute(self, tabuleiro: Tabuleiro):
        contador_seguro = 0

        self.fronteira.put(tabuleiro)
        while True:
            t = self.fronteira.get()
            print(t)
            if t.eSolucao(): return "Solucao encontrada"

            x, y = Tabuleiro.getEspacoEmBranco(t)

            self.adicionarMovimentosAleatorios(t, x, y)

            if contador_seguro > 3:
                print("FINAL")
                for i in list(self.fronteira.queue):
                    print(i)
                return "mais de 10.000 tabuleiros verificados"
            contador_seguro+=1;

    def adicionarMovimentosAleatorios(self, t: Tabuleiro, x, y):
        ordem = np.arange(4)
        random.shuffle(ordem)

        for i in ordem:
            if i == 0 and Movimento.podeMoverCima(y):
                novo = Movimento.moverParaCima(t, x, y)
                self.fronteira.put(novo)
            elif i == 1 and Movimento.podeMoverBaixo(y, t.tamanhoQuadrado):
                novo = Movimento.moverParaBaixo(t, x, y)
                self.fronteira.put(novo)
            elif i == 2 and Movimento.podeMoverEsquerda(x):
                novo = Movimento.moverParaEsquerda(t, x, y)
                self.fronteira.put(novo)
            elif i == 3 and Movimento.podeMoverDireita(x, t.tamanhoQuadrado):
                novo = Movimento.moverParaDireita(t, x, y)
                self.fronteira.put(novo)



def main():
    t = Tabuleiro([[1, 2, 3], [4, 5, 6], [7, 0, 8]])
    x, y = Tabuleiro.getEspacoEmBranco(t)
    b = BuscaEmLargura()
    Movimento.moverParaCima(t,x,y)

    print(b.execute(t))


main()
